File: fast-chat/go/internal/handlers/websocket.go
package handlers

import (
	"encoding/json"
	"log"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gorilla/websocket"
	"github.com/kurze/lab/internal/chat"
)

func createEchoMessage(msg *chat.Message, clientID string) string {
	echoData := chat.MessageData{
		ID:       strconv.FormatInt(msg.ID, 10),
		Nickname: msg.Nickname,
		Text:     msg.Text,
		Time:     msg.Timestamp.Format("15:04:05.000"),
		ClientID: clientID,
		ServerID: strconv.FormatInt(msg.ID, 10),
	}

	echoMsg := chat.ClientMessage{
		Type:      "message",
		Action:    "append",
		Timestamp: time.Now().UnixMilli(),
		Data:      echoData,
	}

	bytes, err := json.Marshal(echoMsg)
	if err != nil {
		return `{"type":"error","data":"Failed to encode echo message"}`
	}
	return string(bytes)
}

func WebSocketHandler(state *chat.ChatState, checkOrigin func(*http.Request) bool, quiet bool) http.HandlerFunc {
	upgrader := websocket.Upgrader{
		ReadBufferSize:  1024,
		WriteBufferSize: 1024,
		CheckOrigin:     checkOrigin,
	}

	return func(w http.ResponseWriter, r *http.Request) {
		ws, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			log.Printf("WebSocket upgrade error: %v", err)
			return
		}

		conn := chat.NewConnection(chat.TransportWebSocket)
		state.AddConnection(conn)

		if !quiet {
			log.Printf("New WebSocket connection: %s", conn.ID)
		}

		go wsWriter(ws, conn)

		wsReader(ws, conn, state, quiet)
	}
}

// wsReader reads messages from the WebSocket
func wsReader(ws *websocket.Conn, conn *chat.Connection, state *chat.ChatState, quiet bool) {
	defer func() {
		// Remove connection and get nickname if this is the first cleanup
		nickname := state.RemoveConnection(conn.ID)
		ws.Close()

		// Only broadcast if we actually removed it (prevents duplicates)
		if nickname != "" {
			// Broadcast user left (JSON)
			sysMsg := chat.SystemMessageJSON(nickname + " left the chat")
			state.Broadcast(sysMsg)

			// Update user count (JSON)
			count := state.ConnectionCount()
			state.Broadcast(chat.UserCountJSON(count))

			if !quiet {
				log.Printf("WebSocket user %s disconnected (%s)", nickname, conn.ID)
			}
		} else {
			if !quiet {
				log.Printf("WebSocket connection already cleaned up: %s", conn.ID)
			}
		}
	}()

	ws.SetReadDeadline(time.Now().Add(60 * time.Second))
	ws.SetPongHandler(func(string) error {
		ws.SetReadDeadline(time.Now().Add(60 * time.Second))
		return nil
	})

	for {
		_, message, err := ws.ReadMessage()
		if err != nil {
			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
				log.Printf("WebSocket error: %v", err)
			}
			break
		}

		conn.UpdateLastSeen()
		handleClientMessage(string(message), conn, state, quiet)
	}
}

// wsWriter writes messages to the WebSocket
func wsWriter(ws *websocket.Conn, conn *chat.Connection) {
	ticker := time.NewTicker(30 * time.Second)
	defer func() {
		ticker.Stop()
		ws.Close()
	}()

	for {
		select {
		case message, ok := <-conn.SendChan:
			if !ok {
				ws.WriteMessage(websocket.CloseMessage, []byte{})
				return
			}

			ws.SetWriteDeadline(time.Now().Add(10 * time.Second))
			if err := ws.WriteMessage(websocket.TextMessage, []byte(message)); err != nil {
				return
			}

		case <-ticker.C:
			ws.SetWriteDeadline(time.Now().Add(10 * time.Second))
			if err := ws.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		}
	}
}

// handleClientMessage processes incoming client messages
func handleClientMessage(data string, conn *chat.Connection, state *chat.ChatState, quiet bool) {
	parts := strings.Split(data, "|")
	if len(parts) == 0 {
		return
	}

	command := parts[0]

	switch command {
	case "JOIN":
		if len(parts) < 2 {
			return
		}
		nickname := parts[1]
		conn.SetNickname(nickname)

		// Broadcast system message (JSON)
		sysMsg := chat.SystemMessageJSON(nickname + " joined the chat")
		state.Broadcast(sysMsg)

		// Update user count (JSON)
		count := state.ConnectionCount()
		state.Broadcast(chat.UserCountJSON(count))

		if !quiet {
			log.Printf("User %s joined (conn: %s)", nickname, conn.ID)
		}

	case "SEND":
		if len(parts) < 3 {
			return
		}
		nickname := parts[1]
		text := parts[2]

		var clientID string
		if len(parts) >= 4 {
			clientID = parts[3]
		}

		msg := state.AddMessage(nickname, text)

		if clientID != "" {
			echoMsg := createEchoMessage(msg, clientID)
			conn.Send(echoMsg)
		}

		jsonMsg := msg.ToJSON()
		state.BroadcastExcept(jsonMsg, conn.ID)

		if !quiet {
			log.Printf("Message from %s: %s", nickname, text)
		}

	case "HISTORY":
		if len(parts) < 2 {
			return
		}

		// Support both old format (HISTORY|count) and new format (HISTORY|skip|take)
		var skip, take int
		var err error

		if len(parts) >= 3 {
			// New format: HISTORY|skip|take
			skip, err = strconv.Atoi(parts[1])
			if err != nil || skip < 0 {
				return
			}
			take, err = strconv.Atoi(parts[2])
			if err != nil || take <= 0 {
				return
			}
		} else {
			// Old format: HISTORY|count (for backward compatibility)
			take, err = strconv.Atoi(parts[1])
			if err != nil || take <= 0 {
				return
			}
			skip = 10 // Skip initial 10
		}

		const maxHistoryTake = 100
		const maxHistorySkip = 10000
		if take > maxHistoryTake {
			take = maxHistoryTake
		}
		if skip > maxHistorySkip {
			skip = maxHistorySkip
		}

		// Get history
		messages := state.GetHistory(skip, take)

		// Send history as JSON with prepend action
		if len(messages) > 0 {
			historyData := make([]chat.MessageData, len(messages))
			for i, msg := range messages {
				historyData[i] = chat.MessageData{
					ID:       strconv.FormatInt(msg.ID, 10),
					Nickname: msg.Nickname,
					Text:     msg.Text,
					Time:     msg.Timestamp.Format("15:04:05.000"),
				}
			}

			historyMsg := chat.ClientMessage{
				Type:      "history",
				Action:    "prepend",
				Timestamp: time.Now().UnixMilli(),
				Data:      historyData,
			}

			jsonBytes, _ := json.Marshal(historyMsg)
			conn.Send(string(jsonBytes))
		}

	case "PING":
		// Keep-alive ping, no response needed
	}
}

File: fast-chat/rust/src/handlers/websocket.rs
use crate::chat::{create_echo_message, history_messages_json, system_message_json, user_count_json, ChatState, Connection, TransportType};
use axum::{
    extract::{
        ws::{Message, WebSocket},
        State, WebSocketUpgrade,
    },
    response::Response,
};
use futures::{sink::SinkExt, stream::StreamExt};
use std::sync::Arc;

pub async fn websocket_handler(
    ws: WebSocketUpgrade,
    State(state): State<Arc<ChatState>>,
) -> Response {
    ws.on_upgrade(|socket| handle_socket(socket, state))
}

async fn handle_socket(socket: WebSocket, state: Arc<ChatState>) {
    let (conn, mut receiver) = Connection::new(TransportType::WebSocket, 256);
    state.add_connection(conn.clone());

    tracing::info!("New WebSocket connection: {}", conn.id);

    let (mut ws_sender, mut ws_receiver) = socket.split();

    let conn_id = conn.id;
    let state_clone = state.clone();
    let send_task = tokio::spawn(async move {
        while let Some(msg) = receiver.recv().await {
            if ws_sender.send(Message::Text(msg)).await.is_err() {
                break;
            }
        }
    });

    while let Some(Ok(msg)) = ws_receiver.next().await {
        if let Message::Text(text) = msg {
            conn.update_last_seen();
            handle_client_message(&text, &conn, &state).await;
        } else if let Message::Close(_) = msg {
            break;
        }
    }

    send_task.abort();

    if let Some(nickname) = state_clone.remove_connection(conn_id) {
        let sys_msg = system_message_json(&format!("{} left the chat", nickname));
        state_clone.broadcast(sys_msg);

        let count = state_clone.connection_count();
        state_clone.broadcast(user_count_json(count));

        tracing::info!("WebSocket user {} disconnected ({})", nickname, conn_id);
    } else {
        tracing::info!("WebSocket connection already cleaned up: {}", conn_id);
    }
}

async fn handle_client_message(data: &str, conn: &Connection, state: &Arc<ChatState>) {
    let parts: Vec<&str> = data.split('|').collect();
    if parts.is_empty() {
        return;
    }

    let command = parts[0];

    match command {
        "JOIN" => {
            if parts.len() < 2 {
                return;
            }
            let nickname = parts[1].to_string();
            conn.set_nickname(nickname.clone());

            let sys_msg = system_message_json(&format!("{} joined the chat", nickname));
            state.broadcast(sys_msg);

            let count = state.connection_count();
            state.broadcast(user_count_json(count));

            tracing::info!("User {} joined (conn: {})", nickname, conn.id);
        }

        "SEND" => {
            if parts.len() < 3 {
                return;
            }
            let nickname = parts[1].to_string();
            let text = parts[2].to_string();
            let client_id = if parts.len() >= 4 {
                Some(parts[3].to_string())
            } else {
                None
            };

            let msg = state.add_message(nickname.clone(), text.clone());

            if let Some(client_id) = client_id {
                let echo_msg = create_echo_message(&msg, client_id);
                conn.send(echo_msg);
            }

            let json_msg = msg.to_json();
            state.broadcast_except(json_msg, conn.id);

            tracing::info!("Message from {}: {}", nickname, text);
        }

        "HISTORY" => {
            if parts.len() < 2 {
                return;
            }

            let (skip, take) = if parts.len() >= 3 {
                let skip = parts[1].parse::<usize>().unwrap_or(0);
                let take = parts[2].parse::<usize>().unwrap_or(10);
                (skip, take)
            } else {
                let take = parts[1].parse::<usize>().unwrap_or(10);
                (10, take)
            };

            let skip = skip.min(10000);
            let take = take.min(100);

            let messages = state.get_history(skip, take);

            if !messages.is_empty() {
                let history_refs: Vec<&crate::chat::Message> = messages.iter().collect();
                let history_json = history_messages_json(history_refs);
                conn.send(history_json);
            }
        }

        "PING" => {
        }

        _ => {}
    }
}
